### PR TITLE
CAMEL-12503: added support for propagating camel headers to kafka and vice versa

### PR DESCRIPTION
**Pull request by tdanylchuk**
Since 0.11.0 Kafka support headers, and it would be awesome to have such feature available also in camel component.

By default it filters camel related headers and propagate filtered kafka headers to camel and vice versa.

### CODE REVIEW

File: components/camel-kafka/src/main/java/org/apache/camel/component/kafka/KafkaConfiguration.java [hunk 1]
```diff
@@ -5,9 +5,9 @@
  * The ASF licenses this file to You under the Apache License, Version 2.0
  * (the "License"); you may not use this file except in compliance with
  * the License.  You may obtain a copy of the License at
- *
- *      http://www.apache.org/licenses/LICENSE-2.0
- *
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
```
Comment (lburgazzoli) [inline components/camel-kafka/src/main/java/org/apache/camel/component/kafka/KafkaConfiguration.java]: don't change the license header, what we expect is https://github.com/apache/camel/blob/master/etc/apache-header.txt

**Comment by oscerd** [comment]
Please raise a JIRA issue and amend the commit message to refer to the issue number

**Comment by tdanylchuk** [comment]
https://issues.apache.org/jira/browse/CAMEL-12503 has been risen 
Commit message includes jira ticket.

File: components/camel-kafka/src/main/java/org/apache/camel/component/kafka/KafkaConsumer.java [hunk 2]
```diff
@@ -423,6 +427,12 @@ public void onPartitionsAssigned(Collection<TopicPartition> partitions) {
         }
     }
 
+    private void propagateHeaders(ConsumerRecord<Object, Object> record, Exchange exchange, HeaderFilterStrategy headerFilterStrategy) {
+        StreamSupport.stream(record.headers().spliterator(), false)
+                .filter(header -> !headerFilterStrategy.applyFilterToCamelHeaders(header.key(), header.value(), exchange))
```
Comment (vasilievip) [inline components/camel-kafka/src/main/java/org/apache/camel/component/kafka/KafkaConsumer.java]: filter(keepIncludedHeaders)?

File: components/camel-kafka/src/main/java/org/apache/camel/component/kafka/KafkaHeaderFilterStrategy.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+import org.apache.camel.impl.DefaultHeaderFilterStrategy;
+
+public class KafkaHeaderFilterStrategy extends DefaultHeaderFilterStrategy {
+
+    public KafkaHeaderFilterStrategy() {
+        initialize();  
+    }
+
+    protected void initialize() {
```
Comment (vasilievip) [inline components/camel-kafka/src/main/java/org/apache/camel/component/kafka/KafkaHeaderFilterStrategy.java]: if you assume super.initialize in ancestors, its better to keep this code in constructor and make initialize abstract and eliminate assumptions

File: components/camel-kafka/src/main/java/org/apache/camel/component/kafka/KafkaProducer.java [hunk 4]
```diff
@@ -194,11 +202,11 @@ public ProducerRecord next() {
                     Object value = tryConvertToSerializedType(exchange, next, endpoint.getConfiguration().getSerializerClass());
 
                     if (hasPartitionKey && hasMessageKey) {
-                        return new ProducerRecord(msgTopic, partitionKey, key, value);
+                        return new ProducerRecord(msgTopic, partitionKey, null, key, value, propagatedHeaders);
```
Comment (vasilievip) [inline components/camel-kafka/src/main/java/org/apache/camel/component/kafka/KafkaProducer.java]: can you replace null with named constants?

File: components/camel-kafka/src/main/java/org/apache/camel/component/kafka/KafkaProducer.java [hunk 5]
```diff
@@ -214,15 +222,39 @@ public void remove() {
...
         } else {
-            record = new ProducerRecord(topic, value);
+            record = new ProducerRecord(topic, null, null, null, value, propagatedHeaders);
         }
         return Collections.singletonList(record).iterator();
     }
 
+    private List<Header> getPropagatedHeaders(Exchange exchange, HeaderFilterStrategy headerFilterStrategy) {
+        return exchange.getIn().getHeaders().entrySet().stream()
+                .filter(entry -> !headerFilterStrategy.applyFilterToExternalHeaders(entry.getKey(), entry.getValue(), exchange))
```
Comment (vasilievip) [inline components/camel-kafka/src/main/java/org/apache/camel/component/kafka/KafkaProducer.java]: filter(includedHeaders)?

File: components/camel-kafka/src/main/java/org/apache/camel/component/kafka/KafkaProducer.java [hunk 6]
```diff
@@ -214,15 +222,39 @@ public void remove() {
...
+                .map(entry -> new RecordHeader(entry.getKey(), getHeaderValue(entry.getValue())))
+                .collect(Collectors.toList());
+    }
+
+    private byte[] getHeaderValue(Object value) {
```
Comment (vasilievip) [inline components/camel-kafka/src/main/java/org/apache/camel/component/kafka/KafkaProducer.java]: Is there particular reason to keep String, Long, integer and byte[]?
can I send Double or Byte or Boolean[]?

Comment (tdanylchuk) [reply components/camel-kafka/src/main/java/org/apache/camel/component/kafka/KafkaProducer.java]: I've put most common types, so if you put unsupported one - value will be null.
The docs of kafka says that we need to bypass `byte[]` , but in camel headers we have `Object`, so I need somehow to know how to convert provided object to bytes.

Although I agree, that `Double` also good to have supported.

File: components/camel-kafka/src/main/java/org/apache/camel/component/kafka/KafkaProducer.java (hunk 4 above)
Comment (tdanylchuk) [reply components/camel-kafka/src/main/java/org/apache/camel/component/kafka/KafkaProducer.java]: a lot of code would be added only for this purpose, is there a need for this?

File: components/camel-kafka/src/main/java/org/apache/camel/component/kafka/KafkaHeaderFilterStrategy.java (hunk 3 above)
Comment (tdanylchuk) [reply components/camel-kafka/src/main/java/org/apache/camel/component/kafka/KafkaHeaderFilterStrategy.java]: wanted to keep convention with other filter strategies
https://github.com/apache/camel/blob/3505e718db48cc0d8be5b47f6c4030c958d93a5f/components/camel-sjms/src/main/java/org/apache/camel/component/sjms/SjmsHeaderFilterStrategy.java
https://github.com/apache/camel/blob/3505e718db48cc0d8be5b47f6c4030c958d93a5f/components/camel-aws/src/main/java/org/apache/camel/component/aws/sqs/SqsHeaderFilterStrategy.java

and other classes...

File: components/camel-kafka/src/main/java/org/apache/camel/component/kafka/KafkaConsumer.java (hunk 2 above)
Comment (tdanylchuk) [reply components/camel-kafka/src/main/java/org/apache/camel/component/kafka/KafkaConsumer.java]: ok, will add additional method for this purpose

File: components/camel-kafka/src/main/java/org/apache/camel/component/kafka/KafkaProducer.java (hunk 5 above)
Comment (tdanylchuk) [reply components/camel-kafka/src/main/java/org/apache/camel/component/kafka/KafkaProducer.java]: will add method

File: components/camel-kafka/src/main/java/org/apache/camel/component/kafka/KafkaProducer.java (hunk 6 above)
Comment (Aloren) [reply components/camel-kafka/src/main/java/org/apache/camel/component/kafka/KafkaProducer.java]: if value type is not supported you just silently put null value -- there is no reason for the dev why this doesn't work: no log or exception... i don't think this is nice

Comment (tdanylchuk) [reply components/camel-kafka/src/main/java/org/apache/camel/component/kafka/KafkaProducer.java]: agree, added `warn` message 
+ added `Double` support

**Comment by oscerd** [comment]
It looks good to me, but I'd like to have feedback from @davsclaus

File: components/camel-kafka/src/main/java/org/apache/camel/component/kafka/KafkaProducer.java (hunk 4 above)
Comment (vasilievip) [reply components/camel-kafka/src/main/java/org/apache/camel/component/kafka/KafkaProducer.java]: yes, this is better than new MyClass(null, null, null, null)...

File: components/camel-kafka/src/main/java/org/apache/camel/component/kafka/KafkaHeaderFilterStrategy.java [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+        // filter out camel id and
+        getInFilter().add("breadcrumbId");
```
Comment (davsclaus) [inline components/camel-kafka/src/main/java/org/apache/camel/component/kafka/KafkaHeaderFilterStrategy.java]: We should include this by default. If you do not want breadcrumbs then you can turn it off on CamelContext.

File: components/camel-kafka/src/main/java/org/apache/camel/component/kafka/KafkaProducer.java [hunk 8]
```diff
@@ -214,15 +223,49 @@ public void remove() {
...
+        } else if (value instanceof Double) {
+            ByteBuffer buffer = ByteBuffer.allocate(Double.BYTES);
+            buffer.putDouble((Double) value);
+            return buffer.array();
+        } else if (value instanceof byte[]) {
+            return (byte[]) value;
+        }
+        log.warn("Cannot propagate header value of type[{}], null will be provided instead. " +
+                "Supported types: String, Integer, Long, Double, byte[].", value != null ? value.getClass() : "null");
+        return null;
```
Comment (davsclaus) [inline components/camel-kafka/src/main/java/org/apache/camel/component/kafka/KafkaProducer.java]: Do you really want a null value for unsupported headers. I think we should drop the header entirely. This is what eg JMS component do and others. Also we should consider avoiding logging noise, eg a WARN log will be for every header. So maybe log at DEBUG by default. And then have better documentation about which header types is supported in the docs.

Comment (tdanylchuk) [reply components/camel-kafka/src/main/java/org/apache/camel/component/kafka/KafkaProducer.java]: agree, keeping the same behaviour is a must, will skip unsupported headers and decrease logging level.

File: components/camel-kafka/src/main/java/org/apache/camel/component/kafka/KafkaHeaderFilterStrategy.java (hunk 7 above)
Comment (tdanylchuk) [reply components/camel-kafka/src/main/java/org/apache/camel/component/kafka/KafkaHeaderFilterStrategy.java]: Thought the purpose of this id is only keeping context linked within single instance.
Should I add additional property for this need or let's rely on headers filter?

Comment (davsclaus) [reply components/camel-kafka/src/main/java/org/apache/camel/component/kafka/KafkaHeaderFilterStrategy.java]: No its also to span over the network etc. So you should basically just remove those code lines.

Comment (tdanylchuk) [reply components/camel-kafka/src/main/java/org/apache/camel/component/kafka/KafkaHeaderFilterStrategy.java]: ok, will be done

**Comment by tdanylchuk** [comment]
Added documentation to `kafka-component.adoc` file, fixed review comments.
@davsclaus please, take a look.

**Pull request merged**